### PR TITLE
3681 solution capabilities epics

### PR DESCRIPTION
**Pull request by RuthBerry96**

### CODE REVIEW

File: src/NHSD.BuyingCatalogue.Solutions.API/Controllers/EpicsController.cs [hunk 1]
```diff
@@ -31,5 +36,21 @@ public ActionResult GetAll([FromRoute][Required]string id)
...
+       // </summary>
+       // <param name="id"></param>
+       // <param name="viewModel"></param>
+       // <returns>A Task representing the operation to update the details of the epics section</returns>
+      [HttpPut]
+      [Route("{id}/sections/epics")]
+      [ProducesResponseType(StatusCodes.Status400BadRequest)]
+      [ProducesResponseType(StatusCodes.Status204NoContent)]
+      [ProducesResponseType(StatusCodes.Status404NotFound)]
+      public async Task<ActionResult> Update([FromRoute] [Required] string id,
```
Comment (sam-roberts-bjss) [inline src/NHSD.BuyingCatalogue.Solutions.API/Controllers/EpicsController.cs]: Keep the pattern of using Async on the end?

File: src/NHSD.BuyingCatalogue.Solutions.API/Controllers/EpicsController.cs [hunk 2]
```diff
@@ -31,5 +36,21 @@ public ActionResult GetAll([FromRoute][Required]string id)
...
+          [FromBody] [Required] UpdateEpicsViewModel viewModel) =>
+          (await _mediator.Send(new UpdateEpicsCommand(id, new HashSet<IClaimedEpic>(viewModel?.ClaimedEpics)))
```
Comment (sam-roberts-bjss) [inline src/NHSD.BuyingCatalogue.Solutions.API/Controllers/EpicsController.cs]: Could we rename the class UpdateEpicsCommand to UpdateClaimedEpicsCommand & follow that pattern for the handler etc..?

File: src/NHSD.BuyingCatalogue.Solutions.Application/Domain/Epics/Epic.cs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+using System;
```
Comment (sam-roberts-bjss) [inline src/NHSD.BuyingCatalogue.Solutions.Application/Domain/Epics/Epic.cs]: Need to review this class. The domain of an epic should look similar to the epic data. Can't remember what this class is trying to represent.

File: src/NHSD.BuyingCatalogue.Solutions.Application/Persistence/Epics/SolutionEpicsUpdater.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using NHSD.BuyingCatalogue.Solutions.Application.Domain.Epics;
+using NHSD.BuyingCatalogue.Solutions.Contracts.Persistence;
+
+namespace NHSD.BuyingCatalogue.Solutions.Application.Persistence.Epics
+{
+    internal sealed class SolutionEpicsUpdater
+    {
+        private readonly ISolutionEpicsRepository _solutionEpicsRepository;
```
Comment (sam-roberts-bjss) [inline src/NHSD.BuyingCatalogue.Solutions.Application/Persistence/Epics/SolutionEpicsUpdater.cs]: Could we change ISolutionEpicsRepository interface (& class) to ISolutionEpicRepository, all the rest of the repositories are singular.

File: tests/NHSD.BuyingCatalogue.Solutions.API.UnitTests/EpicsControllerTests.cs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+                    EpicId = "Epic1",
+                    StatusName = "Passed"
+                }
+            };
+
+            var updateEpicsViewModel = new UpdateEpicsViewModel {ClaimedEpics = claimedEpics};
+
+            var epicsViewModel = new HashSet<IClaimedEpic>(updateEpicsViewModel.ClaimedEpics);
+
+            var viewModel = new UpdateEpicsViewModel {ClaimedEpics = claimedEpics};
```
Comment (sam-roberts-bjss) [inline tests/NHSD.BuyingCatalogue.Solutions.API.UnitTests/EpicsControllerTests.cs]: Duplicate view models, not sure we need both?

File: tests/NHSD.BuyingCatalogue.Solutions.API.UnitTests/EpicsControllerTests.cs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+            _mockMediator
+                .Setup(m => m.Send(It.Is<UpdateClaimedEpicsCommand>(e => e.SolutionId == SolutionId),
+                    It.IsAny<CancellationToken>())).ReturnsAsync(validationModel.Object);
+
+            var result = (await _controller.UpdateAsync(SolutionId, viewModel).ConfigureAwait(false)) as NoContentResult;
+            result.StatusCode.Should().Be((int)HttpStatusCode.NoContent);
+
+            _mockMediator.Verify(
+                m => m.Send(It.Is<UpdateClaimedEpicsCommand>(e => e.SolutionId == SolutionId),
```
Comment (sam-roberts-bjss) [inline tests/NHSD.BuyingCatalogue.Solutions.API.UnitTests/EpicsControllerTests.cs]: Need to come up with a way to ensure the list of epics is what we expect in the verify. One example would be to use an equality comparer or use a call back.

File: tests/NHSD.BuyingCatalogue.Solutions.Application.UnitTests/NHSD.BuyingCatalogue.Solutions.Application.UnitTests.csproj [hunk 7]
```diff
@@ -26,6 +26,7 @@
 
   <ItemGroup>
     <ProjectReference Include="..\..\src\NHSD.BuyingCatalogue.Infrastructure\NHSD.BuyingCatalogue.Infrastructure.csproj" />
+    <ProjectReference Include="..\..\src\NHSD.BuyingCatalogue.Solutions.API\NHSD.BuyingCatalogue.Solutions.API.csproj" />
```
Comment (sam-roberts-bjss) [inline tests/NHSD.BuyingCatalogue.Solutions.Application.UnitTests/NHSD.BuyingCatalogue.Solutions.Application.UnitTests.csproj]: The solution application tests should not know anything about the API layer. Can we check why this reference has been brought in.

File: tests/NHSD.BuyingCatalogue.Solutions.Application.UnitTests/Solutions/UpdateSolutionEpicsTests.cs [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+using NHSD.BuyingCatalogue.Solutions.Application.Commands.UpdateClaimedEpics;
+using NHSD.BuyingCatalogue.Solutions.Application.Commands.Validation;
+using NHSD.BuyingCatalogue.Solutions.Contracts.Epics;
+using NHSD.BuyingCatalogue.Solutions.Contracts.Persistence;
+using NUnit.Framework;
+
+namespace NHSD.BuyingCatalogue.Solutions.Application.UnitTests.Solutions
+{
+    [TestFixture]
+    public sealed class UpdateSolutionEpicsTests
```
Comment (sam-roberts-bjss) [inline tests/NHSD.BuyingCatalogue.Solutions.Application.UnitTests/Solutions/UpdateSolutionEpicsTests.cs]: Rename to UpdateClaimedEpicsHandlerTests?

File: tests/NHSD.BuyingCatalogue.Solutions.Application.UnitTests/Solutions/UpdateSolutionEpicsTests.cs [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    {
+        private TestContext _context;
+
+        private const string ValidSolutionId = "Sln1";
+        private const string InvalidSolutionId = "Sln123";
+
+
```
Comment (sam-roberts-bjss) [inline tests/NHSD.BuyingCatalogue.Solutions.Application.UnitTests/Solutions/UpdateSolutionEpicsTests.cs]: Extra new line?

File: tests/NHSD.BuyingCatalogue.Solutions.Application.UnitTests/Solutions/UpdateSolutionEpicsTests.cs [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+                .ReturnsAsync(true);
+            _context.MockSolutionRepository.Setup(x => x.CheckExists(InvalidSolutionId, It.IsAny<CancellationToken>()))
+                .ReturnsAsync(false);
+
+        }
+
+        [Test]
+        public async Task ShouldUpdateEpicsAsync()
+        {
+            var epic = new ClaimedEpicViewModel
```
Comment (sam-roberts-bjss) [inline tests/NHSD.BuyingCatalogue.Solutions.Application.UnitTests/Solutions/UpdateSolutionEpicsTests.cs]: Not sure why we have a reference to the ClaimedEpicViewModel type

File: tests/NHSD.BuyingCatalogue.Solutions.Application.UnitTests/Solutions/UpdateSolutionEpicsTests.cs [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+            Assert.ThrowsAsync<NotFoundException>(() => UpdateEpicsAsync(InvalidSolutionId, claimedEpics));
+        }
+
+        [Test]
+        public void ShouldThrowNotFoundExceptionWhenSolutionIsNotFound()
+        {
+            Assert.ThrowsAsync<NotFoundException>(() => UpdateEpicsAsync(InvalidSolutionId, new HashSet<IClaimedEpic>()));
+        }
+
+        private async Task<ISimpleResult> UpdateEpicsAsync(string solutionId, HashSet<IClaimedEpic> claimedEpics)
```
Comment (sam-roberts-bjss) [inline tests/NHSD.BuyingCatalogue.Solutions.Application.UnitTests/Solutions/UpdateSolutionEpicsTests.cs]: UpdateClaimedEpicsAsync?

File: tests/NHSD.BuyingCatalogue.Solutions.Application.UnitTests/TestContext.cs [hunk 12]
```diff
@@ -336,7 +345,8 @@ private class Scope
                 IRequestHandler<UpdateSupplierCommand, ISimpleResult> updateSupplierRequestHandler,
                 IRequestHandler<UpdateIntegrationsCommand, ISimpleResult> updateIntegrationsRequestHandler,
                 IRequestHandler<UpdateImplementationTimescalesCommand, ISimpleResult> updateImplementationTimescalesHandler,
-                IRequestHandler<UpdateCapabilitiesCommand, ISimpleResult> updateCapabilitiesHandler)
+                IRequestHandler<UpdateCapabilitiesCommand, ISimpleResult> updateCapabilitiesHandler,
+                IRequestHandler<UpdateClaimedEpicsCommand, ISimpleResult> updateEpicsHandler)
```
Comment (sam-roberts-bjss) [inline tests/NHSD.BuyingCatalogue.Solutions.Application.UnitTests/TestContext.cs]: Variable name -> updateClaimedEpicsHandler?

File: tests/NHSD.BuyingCatalogue.Solutions.Persistence.DatabaseTests/SolutionEpicRepositoryTests.cs [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+            CreateCapability("Cap2", "Desc2", "Ref2","1.0.0","http://cap2.link")
+        };
+
+        private readonly List<EpicDetails> _epicDetails = new List<EpicDetails>()
+        {
+            CreateEpic("Epic1", "Name1", _capDetails[0].Id),
+            CreateEpic("Epic2", "Name2", _capDetails[1].Id)
+        };
+
+        private static CapabilityDetails CreateCapability(string name, string desc, string reference, string version, string sourceUrl)
```
Comment (sam-roberts-bjss) [inline tests/NHSD.BuyingCatalogue.Solutions.Persistence.DatabaseTests/SolutionEpicRepositoryTests.cs]: Could we introduce an CapabilityDetailsBuilder similar to the other builders but just for test object creation?

File: tests/NHSD.BuyingCatalogue.Solutions.Persistence.DatabaseTests/SolutionEpicRepositoryTests.cs [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+                Id = Guid.NewGuid(),
+                Name = name,
+                Desc = desc,
+                Reference = reference,
+                Version = version,
+                SourceUrl = sourceUrl
+            };
+        }
+
+        private static EpicDetails CreateEpic(string id, string name, Guid capId)
```
Comment (sam-roberts-bjss) [inline tests/NHSD.BuyingCatalogue.Solutions.Persistence.DatabaseTests/SolutionEpicRepositoryTests.cs]: Same here -> Look to use the builder pattern?

File: tests/NHSD.BuyingCatalogue.Solutions.Persistence.DatabaseTests/SolutionEpicRepositoryTests.cs [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+
+            TestContext testContext = new TestContext();
+            _solutionEpicRepository = testContext.SolutionEpicRepository;
+        }
+
+        [Test]
+        public async Task UpdateSolutionWithOneEpicAsync()
+        {
+            await InsertCapabilityAsync(_capDetails[0]).ConfigureAwait(false);
+            await InsertSolutionCapabilityAsync(Solution1Id, _capDetails[0].Id).ConfigureAwait(false);
```
Comment (sam-roberts-bjss) [inline tests/NHSD.BuyingCatalogue.Solutions.Persistence.DatabaseTests/SolutionEpicRepositoryTests.cs]: Can whether or not we need to insert the solution capability here?

File: tests/NHSD.BuyingCatalogue.Solutions.Persistence.DatabaseTests/SolutionEpicRepositoryTests.cs [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+            await InsertEpicAsync(_epicDetails[0]).ConfigureAwait(false);
+
+            var expectedEpic = new List<IClaimedEpicResult>
```
Comment (sam-roberts-bjss) [inline tests/NHSD.BuyingCatalogue.Solutions.Persistence.DatabaseTests/SolutionEpicRepositoryTests.cs]: Variable name -> expectedClaimedEpics?

File: tests/NHSD.BuyingCatalogue.Solutions.Persistence.DatabaseTests/SolutionEpicRepositoryTests.cs [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+            {
+                Mock.Of<IClaimedEpicResult>(e => e.EpicId == _epicDetails[0].Id && e.StatusName == StatusPassed),
+                Mock.Of<IClaimedEpicResult>(e => e.EpicId == _epicDetails[1].Id && e.StatusName == StatusPassed)
+            };
+
+            await _solutionEpicRepository
+                .UpdateSolutionEpicAsync(Solution1Id,
+                    Mock.Of<IUpdateClaimedRequest>(c => c.ClaimedEpics == expectedEpic), new CancellationToken()).ConfigureAwait(false);
```
Comment (sam-roberts-bjss) [inline tests/NHSD.BuyingCatalogue.Solutions.Persistence.DatabaseTests/SolutionEpicRepositoryTests.cs]: Just noticed this interface name, could we change it to IUpdateClaimedEpicListRequest?

File: tests/NHSD.BuyingCatalogue.Testing.Data/Entities/SolutionEpicEntity.cs [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+        (
+            @SolutionId
+           ,@CapabilityId
+           ,@EpicId
+           ,@StatusId
+           ,@LastUpdated
+           ,@LastUpdatedBy
+        )";
+
+        public static async Task<IEnumerable<string>> FetchForSolutionAsync(string solutionId)
```
Comment (sam-roberts-bjss) [inline tests/NHSD.BuyingCatalogue.Testing.Data/Entities/SolutionEpicEntity.cs]: Rename method -> FetchAllEpicIdsForSolutionAsync?

File: tests/NHSD.BuyingCatalogue.Testing.Data/EntityBuilders/SolutionEpicEntityBuilder.cs [hunk 19]
```diff
@@ -0,0 +1,56 @@
+using System;
+using NHSD.BuyingCatalogue.Testing.Data.Entities;
+
+namespace NHSD.BuyingCatalogue.Testing.Data.EntityBuilders
+{
+    public sealed class SolutionEpicEntityBuilder
+    {
+        private readonly SolutionEpicEntity _solutionEpicEntity;
+        public static SolutionEpicEntityBuilder Create()
```
Comment (sam-roberts-bjss) [inline tests/NHSD.BuyingCatalogue.Testing.Data/EntityBuilders/SolutionEpicEntityBuilder.cs]: Add new line?

File: tests/NHSD.BuyingCatalogue.Testing.Data/EntityBuilders/SolutionEpicEntityBuilder.cs [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+        public SolutionEpicEntityBuilder()
+        {
+            //Default
+            _solutionEpicEntity = new SolutionEpicEntity()
+            {
+                SolutionId = "SolutionId",
+                CapabilityId = Guid.NewGuid(),
+                EpicId = "EpicId",
+                
+                StatusId = 1
```
Comment (sam-roberts-bjss) [inline tests/NHSD.BuyingCatalogue.Testing.Data/EntityBuilders/SolutionEpicEntityBuilder.cs]: Constant for the value 1 -> Not sure what 1 means

File: tests/NHSD.BuyingCatalogue.Solutions.API.UnitTests/EpicsControllerTests.cs (hunk 6 above)
Comment (RuthBerry96) [reply tests/NHSD.BuyingCatalogue.Solutions.API.UnitTests/EpicsControllerTests.cs]: Task created in the backlog #5069

File: src/NHSD.BuyingCatalogue.Solutions.Application/Persistence/Epics/UpdateClaimedEpicRequest.cs [hunk 21]
```diff
@@ -4,7 +4,7 @@
 
 namespace NHSD.BuyingCatalogue.Solutions.Application.Persistence.Epics
 {
-    internal sealed class UpdateClaimedEpicRequest : IUpdateClaimedRequest
+    internal sealed class UpdateClaimedEpicRequest : IUpdateClaimedEpicListRequest
```
Comment (sam-roberts-bjss) [inline src/NHSD.BuyingCatalogue.Solutions.Application/Persistence/Epics/UpdateClaimedEpicRequest.cs]: Can the class name match the interface name?

File: tests/NHSD.BuyingCatalogue.Testing.Data/EntityBuilders/SolutionEpicEntityBuilder.cs [hunk 22]
```diff
@@ -6,6 +6,8 @@ namespace NHSD.BuyingCatalogue.Testing.Data.EntityBuilders
     public sealed class SolutionEpicEntityBuilder
     {
         private readonly SolutionEpicEntity _solutionEpicEntity;
+        private const int StatusId = 1;
```
Comment (sam-roberts-bjss) [inline tests/NHSD.BuyingCatalogue.Testing.Data/EntityBuilders/SolutionEpicEntityBuilder.cs]: I was thinking here that the name of the constant could provide a bit more meaning. What does a status of 1 refer to?

**Pull request merged**